### PR TITLE
Ports 'Save 0.6-0.7s of init time by splitting registering lists of signals into its own proc' and micro-optimizes the SEND_SIGNAL() define

### PR DESCRIPTION
**Pull request by Tsar-Salat**
Ports:
- https://github.com/tgstation/tgstation/pull/61319
- https://github.com/tgstation/tgstation/pull/71056

## About The Pull Request

Signal Optimizations for faster init, supposedly

PR 1

if(list?[index]) is very slightly faster than if(list && list[index]), so we converts the SEND_SIGNAL() define to that. 

PR 2

> - Makes QDELETED use isnull(x) instead of !x, giving about 0.2 to 0.25s of speed.
> - Make disposal constructs only update icon state rather than go through expensive overlay code. Unfortunately did not have - much effect, but is something they should've been doing nonetheless.
> - Makes RegisterSignal only take signals directly as opposed to allocating a fresh list of signals. Very few consumers actually used this and it costs about 0.4s. Also I think this is just a bad API anyway and that separate procs are important
> 

## Why It's Good For The Game

Faster init

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image

</details>

## Changelog
:cl: rkz, Mothblocks, Kylerace
refactor: speeds up init by ~0.8s by splitting registering lists of signals
code: documents component code
/:cl:

### CODE REVIEW

File: code/datums/components/plumbing/_plumbing.dm [hunk 1]
```diff
@@ -27,10 +27,10 @@
...
 	turn_connects = _turn_connects
 
-	RegisterSignal(parent, list(COMSIG_MOVABLE_MOVED,COMSIG_PARENT_PREQDELETED), PROC_REF(disable))
-	RegisterSignal(parent, list(COMSIG_OBJ_DEFAULT_UNFASTEN_WRENCH), PROC_REF(toggle_active))
-	RegisterSignal(parent, list(COMSIG_OBJ_HIDE), PROC_REF(hide))
-	RegisterSignal(parent, list(COMSIG_ATOM_UPDATE_OVERLAYS), PROC_REF(create_overlays)) //create overlays also gets called after init (no idea by what it just happens)
+	RegisterSignals(parent, list(COMSIG_MOVABLE_MOVED,COMSIG_PARENT_PREQDELETED), PROC_REF(disable))
+	RegisterSignals(parent, list(COMSIG_OBJ_DEFAULT_UNFASTEN_WRENCH), PROC_REF(toggle_active))
+	RegisterSignals(parent, COMSIG_OBJ_HIDE, PROC_REF(hide))
+	RegisterSignals(parent, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(create_overlays)) //create overlays also gets called after init (no idea by what it just happens)
```
Comment (EvilDragonfiend) [inline code/datums/components/plumbing/_plumbing.dm]: this is wrong

File: code/datums/components/_component.dm [hunk 2]
```diff
@@ -52,70 +117,121 @@
...
+		lookup[signal_type] = src
+	else if(looked_up == src) // We already registered here
+		return
+	else if(!length(looked_up)) // One other thing registered here
+		lookup[signal_type] = list((looked_up) = TRUE, (src) = TRUE)
+	else // Many other things have registered here
+		looked_up[src] = TRUE
+
+/// Registers multiple signals to the same proc.
+/datum/proc/RegisterSignals(datum/target, list/signal_types, proctype, override = FALSE)
```
Comment (EvilDragonfiend) [inline code/datums/components/_component.dm:220]: I think it needs this:
```dm
if(!islist(signal_types))
	stack_trace("[target] called RegisterSignals() with non-list signal: [signal_types]")
	RegisterSignal(target, signal_types, proctype, override)
	return
```

**Review by EvilDragonfiend — DISMISSED** [review]
Looks good to me except there are a few things:
* RegisterSignals should get a failsafe when it took a non-list, since there was a mistake, we need to make it sure.
* There are a few register signals doing with a single item in a list. I think it doesn't have to be "signals" and "list", but well, it's a port anyway, optional.

File: code/datums/components/plumbing/_plumbing.dm (hunk 1 above)
Comment (Tsar-Salat) [reply code/datums/components/plumbing/_plumbing.dm]: im dumb

Comment (PowerfulBacon) [reply code/datums/components/plumbing/_plumbing.dm]: It would be nice if we could lint against this

**Review by itsmeow — APPROVED** [review]

**Pull request merged**